app/src/main/java/com/arctouch/codechallenge/ui/activity/presenter/implementation/HomeActivityPresenterImplementation.java

```diff
@@ -0,0 +1,131 @@
+package com.arctouch.codechallenge.ui.activity.presenter.implementation;
+
+import android.util.Log;
+
+import com.arctouch.codechallenge.callback.OnGetGenres;
+import com.arctouch.codechallenge.callback.OnGetSearchMovies;
+import com.arctouch.codechallenge.callback.OnGetUpcomingMovies;
+import com.arctouch.codechallenge.data.Cache;
+import com.arctouch.codechallenge.entity.Genre;
+import com.arctouch.codechallenge.entity.Movie;
+import com.arctouch.codechallenge.entity.UpcomingMoviesResponse;
+import com.arctouch.codechallenge.repository.GenresRepository;
+import com.arctouch.codechallenge.repository.SearchMoviesRepository;
+import com.arctouch.codechallenge.repository.UpcomingMoviesRepository;
+import com.arctouch.codechallenge.repository.implementation.GenresRepositoryImplementation;
+import com.arctouch.codechallenge.repository.implementation.SearchMoviesRepositoryImplementation;
+import com.arctouch.codechallenge.repository.implementation.UpcomingMoviesRepositoryImplementation;
+import com.arctouch.codechallenge.ui.activity.presenter.HomeActivityPresenter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class HomeActivityPresenterImplementation implements HomeActivityPresenter, OnGetUpcomingMovies, OnGetGenres, OnGetSearchMovies{
+
+    private static final String TAG = "HomeActivity Pres.";
+
+    private HomeActivityPresenter.View mView;
+    private int mPage = 0;
+
+
+    private UpcomingMoviesRepository upcomingMoviesRepository;
+    private GenresRepository genresRepository;
+    private SearchMoviesRepository searchMoviesRepository;
+
+    public HomeActivityPresenterImplementation(View mView) {
+        upcomingMoviesRepository = new UpcomingMoviesRepositoryImplementation();
+        genresRepository = new GenresRepositoryImplementation();
+        searchMoviesRepository = new SearchMoviesRepositoryImplementation();
+        this.mView = mView;
+    }
+
+    @Override
+    public void getUpcomingMovies() {
+        upcomingMoviesRepository.getUpcomingMovies(this);
+    }
+
+    @Override
+    public void getMoreUpcomingMovies() {
+        upcomingMoviesRepository.getUpcomingMovies(this.mPage, this);
+    }
+
+    @Override
+    public void start() {
+        //get genres first to update cache
+        genresRepository.getGenres(this);
```

Comment:
You could have added some network connection handling

Reply:
True! That and also some feedbacks to the user in case something goes wrong, by implementing error messages and actually using them. On submitted code I only log errors.